libbpf-tools/klockstat.bpf.c

```diff
@@ -0,0 +1,265 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 Google LLC.
+ *
+ * Based on klockstat from BCC by Jiri Olsa and others
+ * 2021-10-26   Barret Rhoden   Created this.
+ */
+#include "vmlinux.h"
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "klockstat.h"
+#include "bits.bpf.h"
+
+const volatile pid_t targ_tgid = 0;
+const volatile pid_t targ_pid = 0;
+struct mutex *const volatile targ_lock = NULL;
+
+/* Make sure all programs are loaded before we start tracing */
+bool enabled = false;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(max_entries, MAX_ENTRIES);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, PERF_MAX_STACK_DEPTH * sizeof(u64));
+} stack_map SEC(".maps");
+
+/* depth == nr of locks held, including the currently traced lock. */
+struct task_info {
+	u32 depth;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, struct task_info);
+} task_map SEC(".maps");
+
+/* Represents a task at a given lock depth, used for lookups. */
+struct depth_id {
+	u64 task_id;
+	u32 depth;
+};
+
+struct lockholder_info {
+	s32 stack_id;
+	u64 task_id;
+	u64 try_at;
+	u64 acq_at;
+	u64 rel_at;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct depth_id);
+	__type(value, struct lockholder_info);
+} lockholder_map SEC(".maps");
+
+/*
+ * Keyed by stack_id.
+ *
+ * Multiple call sites may have the same underlying lock, but we only know the
+ * stats for a particular stack frame.  Multiple tasks may have the same
+ * stackframe.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, s32);
+	__type(value, struct lock_stat);
+} stat_map SEC(".maps");
+
+static bool tracing_task(u64 task_id)
+{
+	u32 tgid = task_id >> 32;
+	u32 pid = task_id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return false;
+	if (targ_pid && targ_pid != pid)
+		return false;
+	return true;
+}
+
+static void lock_contended(void *ctx, struct mutex *lock)
+{
+	u64 task_id;
+	struct task_info *ti;
+	struct lockholder_info li[1] = {0};
+	struct depth_id di[1] = {0};
```

Comment:
Don't think you need this single element array, `struct foo bar = {}` should be enough.

Reply:
i do that so you can use the variable as a pointer.  otherwise you'd have 'li' being accessed with '.' in some functions and '->' in others.

though di is always local to the function, so i changed that.

i can change it for 'li' too if you want.

Reply:
It's ok. I don't have strong opinion about this.
Someone does this before, see: https://github.com/iovisor/bcc/blob/4d08181d194bc2961f5d3e599f6db5d870648d15/libbpf-tools/uprobe_helpers.c#L232-L234